scipy/spatial/_kdtree.py

```diff
@@ -454,31 +448,7 @@ def query(
             raise TypeError("KDTree does not work with complex data")
 
         if k is None:
-            # k=None, return all neighbors
-            warnings.warn(
-                "KDTree.query with k=None is deprecated and will be removed "
-                "in SciPy 1.8.0. Use KDTree.query_ball_point instead.",
-                DeprecationWarning)
-
-            # Convert index query to a lists of distance and index,
-            # sorted by distance
-            def inds_to_hits(point, neighbors):
-                dist = minkowski_distance(point, self.data[neighbors], p)
-                hits = sorted([(d, i) for d, i in zip(dist, neighbors)])
-                return [d for d, i in hits], [i for d, i in hits]
-
-            x = np.asarray(x, dtype=np.float64)
-            inds = super().query_ball_point(
-                x, distance_upper_bound, p, eps, workers)
-
-            if isinstance(inds, list):
-                return inds_to_hits(x, inds)
-
-            dists = np.empty_like(inds)
-            for idx in np.ndindex(inds.shape):
-                dists[idx], inds[idx] = inds_to_hits(x[idx], inds[idx])
-
-            return dists, inds
+            raise TypeError("k must be an integer or a sequence of integers")
```

Comment:
If we want to add a validation, I would rather add a real one which is not just checking for None.

Otherwise, we tend to use `ValidationError` for such things.
```suggestion
            raise ValueError("k must be an integer or a sequence of integers")
```